go/nbs/table_set.go

```diff
@@ -5,165 +5,281 @@
-		reads += rds
-		if !remaining {
-			return reads, split, remaining
+func (ts tableSet) calcReads(reqs []getRecord, blockSize uint64) (reads int, split, remaining bool) {
+	f := func(css chunkSources) (reads int, split, remaining bool) {
+		for _, haver := range css {
+			rds, remaining := haver.calcReads(reqs, blockSize)
+			reads += rds
+			if !remaining {
+				return reads, split, remaining
+			}
+			split = true
 		}
-		split = true
+		return reads, split, remaining
+	}
+	reads, split, remaining = f(ts.novel)
+	if remaining {
+		var rds int
+		rds, split, remaining = f(ts.upstream)
+		reads += rds
 	}
 	return reads, split, remaining
 }
 
-func (css chunkSources) count() (count uint32) {
-	for _, haver := range css {
-		count += haver.count()
+func (ts tableSet) count() (count uint32) {
+	f := func(css chunkSources) (count uint32) {
+		for _, haver := range css {
+			count += haver.count()
+		}
+		return
 	}
-	return
+	return f(ts.novel) + f(ts.upstream)
+}
+
+func (ts tableSet) data() (data uint64) {
+	f := func(css chunkSources) (data uint64) {
+		for _, haver := range css {
+			data += haver.data()
+		}
+		return
+	}
+	return f(ts.novel) + f(ts.upstream)
+}
+
+// Size returns the number of tables in this tableSet.
+func (ts tableSet) Size() int {
+	return len(ts.novel) + len(ts.upstream)
 }
 
 // Prepend adds a memTable to an existing tableSet, compacting |mt| and
 // returning a new tableSet with newly compacted table added.
 func (ts tableSet) Prepend(mt *memTable) tableSet {
-	newTables := make(chunkSources, len(ts.chunkSources)+1)
-	newTables[0] = newCompactingChunkSource(mt, ts, ts.p, ts.rl)
-	copy(newTables[1:], ts.chunkSources)
-	return tableSet{newTables, ts.p, ts.rl}
+	newTs := tableSet{
+		novel:    make(chunkSources, len(ts.novel)+1),
+		upstream: make(chunkSources, len(ts.upstream)),
+		p:        ts.p,
+		rl:       ts.rl,
+	}
+	newTs.novel[0] = newCompactingChunkSource(mt, ts, ts.p, ts.rl)
+	copy(newTs.novel[1:], ts.novel)
+	copy(newTs.upstream, ts.upstream)
+	return newTs
+}
+
+// Compact returns a new tableSet that's smaller than |ts|. It takes the
+// max(2, len(ts)/2) smallest upstream tables, compacts them into a single
+// large table, then returns a new tableSet with novel = ts.novel, and
+// upstream set to this new large table and the not-compacted members of
+// ts.upstream. The compactees are returned separately so that the caller can
+// close them if she so chooses.
+func (ts tableSet) Compact() (ns tableSet, compactees chunkSources) {
+	ns = tableSet{
+		novel:    make(chunkSources, len(ts.novel)),
+		upstream: make(chunkSources, len(ts.upstream), len(ts.upstream)+1),
+		p:        ts.p,
+		rl:       ts.rl,
+	}
+	copy(ns.novel, ts.novel)
+	copy(ns.upstream, ts.upstream)
+	sort.Sort(chunkSourcesByDescendingCount(ns.upstream))
+
+	max := func(a, b int) int {
+		if a > b {
+			return a
+		}
+		return b
+	}
+
+	partition := len(ns.upstream) - max(2, len(ns.upstream)/2)
```

Comment:
I think the fanciness here in order to avoid allocating a new slice buffer isn't worth it. can you just do the naive thing here that's easy to read. i.e.

```
tablesToCompact := ...
compactedTable := ...
upstream := 
```